web/src/repo/LSIFVerification.tsx

```
@@ -0,0 +1,110 @@
+                            the following name:
+                        </div>
+
+                        {this.state.challenge ? (
+                            <>
+                                <CopyableText text={this.state.challenge} size={16} />
+                                <button
+                                    className="btn btn-primary"
+                                    disabled={this.state.verifying}
+                                    onClick={this.verify}
+                                >
+                                    {this.state.verifying && <LoadingSpinner className="icon-inline" />}
+                                    Check now
+                                </button>
+                                {isErrorLike(this.state.tokenOrError) && (
+                                    <div className="alert alert-danger">{this.state.tokenOrError.toString()}</div>
+                                )}
+                            </>
+                        ) : (
+                            <div>Loading challenge...</div>
+                        )}
+                    </>
+                )}
+            </div>
+        )
+    }
+
+    private verify = async () => {
+        this.setState({ tokenOrError: undefined, verifying: true })
+        try {
+            const url = new URL('/.api/lsif/verify', window.location.href)
+            url.searchParams.set('repository', this.props.repoName)
+            const response: VerifyResponse = await (await fetch(url.href, {
+                headers: {
+                    'X-Requested-With': 'Sourcegraph',
+                },
+            })).json()
+            if ('Failure' in response) {
+                throw new Error(response.Failure)
+            }
+            this.setState({ tokenOrError: response.Token })
+        } catch (error) {
+            this.setState({ tokenOrError: error })
+        } finally {
+            this.setState({ verifying: false })
+        }
+    }
```

I see a lot of problems with the async code in this component. I am confused why you chose to use async functions + fetch instead of the RxJS patterns we use in every other component, which amount to the same code but would avoid all these problems:

- HTTP status codes are not checked for
- `fetchChallenge()` errors are not handled
- `setState()` can be called after component is unmounted, which makes React throw
- Race conditions from triggering multiple verifications at the same time because no cancellation

I understand that this is just for GopherCon, but the equivalent Rx code is what we use everywhere else and I would expect it to not take any longer to write. In the end, even for GopherCon, I think we should present people with the best possible experience with this demo, because nothing sucks more than demos not working properly. Since this is not behind a feature flag, these errors could impact people just using Sourcegraph normally too.

---

What RxJS patterns are you referring to? Putting a `Subscription` between the `fetch` and `setState()` to avoid calling `setState()` after the component is unmounted? That didn't occur to me earlier.

I'll wait for your response before changing this.

---

Using RxJS in general (`ajax()`, `Subscription`, etc) in this component instead of async functions and `fetch()`, like in any other React component that does HTTP requests.